### PR TITLE
Modification of 4_SparseKernelMethods.ipynb and 5_CUR.ipynb 

### PR DESCRIPTION
**Pull request by PKGuo**
Fixed notation errors to resolve mismatches between formulas / literal expressions and their code implementations

Added some details to help readers understand.

### CODE REVIEW

**Comment by rosecers** [comment]
@PKGuo You can explicitly make this a "Draft PR"

https://github.blog/2019-02-14-introducing-draft-pull-requests/

File: notebooks/4_SparseKernelMethods.ipynb [hunk 1]
```diff
@@ -302,20 +302,25 @@
...
    "metadata": {},
    "source": [
-    "$\\mathbf{\\tilde{\\Phi}}$ is the feature matrix for the train points in the RKHS defined by the $M$ active set. Sparse KPCA can be understood (and derived) as PCA in the active set RKHS, by computing and diagonalising the covariance matrix built from $\\mathbf{\\Phi}_{NM}$. The covariance should be computed using *centered* kernel features, as discussed above\n",
+    "$\\mathbf{\\tilde{\\Phi}}$ is the feature matrix for the train points in the RKHS defined by the $M$ active set. Sparse KPCA can be understood (and derived) as PCA in the active set RKHS, by computing and diagonalising the covariance matrix built from $\\mathbf{\\tilde{\\Phi}}$. The covariance should be computed using *centered* kernel features, as discussed above\n",
     "\n",
     "\\begin{equation}\n",
     "\\mathbf{C} = \\mathbf{\\tilde{\\Phi}}^T \\mathbf{\\tilde{\\Phi}} = \\mathbf{U}_C \\mathbf{\\Lambda}_C \\mathbf{U}_C^T.\n",
     "\\end{equation}\n",
     "\n",
+    "<!---\n",
```
Comment (rosecers) [inline notebooks/4_SparseKernelMethods.ipynb]: Please don't include comments in PRs

**Comment by rosecers** [comment]
@PKGuo with a rebase this should be ready for merge.

File: notebooks/4_SparseKernelMethods.ipynb [hunk 2]
```diff
@@ -267,7 +267,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "vmm, Umm = sorted_eig(Kmm, thresh=1e-12)\n",
+    "vmm, Umm = sorted_eig(Kmm, thresh=0)\n",
     "\n",
     "Phi = np.matmul(Knm_train, Umm[:,:n_active-1])\n",
     "Phi = np.matmul(Phi, np.diagflat(1.0/np.sqrt(vmm[0:n_active-1])))"
```
Comment (bhelfrecht) [inline notebooks/4_SparseKernelMethods.ipynb:273]: If `sorted_eig` is already discarding small eigenvalues, do we still need all the `n_active-1` stuff?
@rosecers why does the `sorted_eig` function set small eigenvalues to zero in some cases and just discard them in others?

File: notebooks/4_SparseKernelMethods.ipynb [hunk 3]
```diff
@@ -648,7 +636,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "For Sparse KPCovR, instead of using the Nystr&ouml;m approximation as in previous sparse methods, we formulate sparse KPCovR from KPCovR in a similar manner to how we derived feature space PCovR from structure space PCovR in the [PCovR Notebook](2_PrincipalCovariatesRegression.ipynb).\n",
+    "For Sparse KPCovR, instead of using the Nystr&ouml;m approximation as in previous sparse methods, we formulate sparse KPCovR from KPCovR in a similar manner to how we derived feature space PCovR from sample space PCovR in the [PCovR Notebook](2_PrincipalCovariatesRegression.ipynb).\n",
```
Comment (bhelfrecht) [inline notebooks/4_SparseKernelMethods.ipynb:639]: Sparse KPCovR is actually based on _feature space_ PCovR; regular KPCovR is based on _sample space_ PCovR

File: notebooks/5_CUR.ipynb [hunk 4]
```diff
@@ -162,10 +162,10 @@
     "For the $j^{th}$ column, the importance score is given by\n",
     "\n",
     "\\begin{equation}\n",
-    "\\pi_j = \\sum_i^k (U_C)_{ij}^2,\n",
+    "\\pi_j = \\sum_i^k (U_C)_{ji}^2,\n",
```
Comment (bhelfrecht) [inline notebooks/5_CUR.ipynb]: Aren't the columns of Uc still the eigenvectors? So we should still be summing over the rows, i.e., (U_C)_{ij}, correct? The inconsistency with the code comes in where the SVD returns the right singular vectors transposed, so that the eigenvectors are rows. So we should instead probably revert this change and rename the variable `U_C` to something like `U_CT` an add a code comment to this effect

File: notebooks/5_CUR.ipynb [hunk 5]
```diff
@@ -162,10 +162,10 @@
...
     "\\end{equation}\n",
     "\n",
-    "where $k$ is again the target rank of the new matrix and $(U_C)_{ij}$ is the value in the $i^{th}$ row and $j^{th}$ column of $\\mathbf{U}_C$."
+    "where $k$ is again the target rank of the new matrix and $(U_C)_{ji}$ is the value in the $j^{th}$ row and $i^{th}$ column of $\\mathbf{U}_C$."
    ]
```
Comment (bhelfrecht) [inline notebooks/5_CUR.ipynb]: See comment above; I think this change should probably be reverted

**Review by bhelfrecht — CHANGES_REQUESTED** [review]

File: notebooks/4_SparseKernelMethods.ipynb (hunk 3 above)
Comment (PKGuo) [reply notebooks/4_SparseKernelMethods.ipynb:639]: I don't get your point. I think what this sentence means is that we derive the formula of sparse KPCovR in **a similar way** to the derivation of the formula of regular feature space PCovR.

We derive the formula of ****feature** space** PCovR from ****sample** space** PCovR.

image
image


Then we derive the formula of sparse KPCovR in a similar way.
image

File: notebooks/4_SparseKernelMethods.ipynb (hunk 2 above)
Comment (PKGuo) [reply notebooks/4_SparseKernelMethods.ipynb:273]: I also think n_active-1 is meaningless. Here, I follow @rosecers 's suggestion to set the threshold to 0, and then use n_active-1 to remove zero eigenvalues.

File: notebooks/5_CUR.ipynb (hunk 4 above)
Comment (PKGuo) [reply notebooks/5_CUR.ipynb]: Here I use the formula mentioned in a [reference](https://doi.org/10.1063/1.5024611) for the fourth notebook. If we want to know the importance score of the j-th column in X, we sum the j-th coordinate of different columns (right singular vectors) in U_C, i.e., the first k coordinates of the j-th row in U_C.
image

@rosecers sent me your new PCov-CUR paper. But its explanation of the summation process is not as detailed as the previous one. I don't quiet understand it.
image

Comment (PKGuo) [reply notebooks/5_CUR.ipynb]: In addition, the importance scoring formula I have seen in other papers is the same as in the first paper. They all sum the j-th coordinates of different columns in UC.
image

image

Comment (PKGuo) [reply notebooks/5_CUR.ipynb]: I once added a note to remind readers that sps.linalg.svds will return U_C^T containing right singular vectors as rows. So we have two different ways of summing before and after. But Rose said it would be unnecessary after a new PR. So I deleted it.
image

image
**Note** that the Ct in U_Ct here refers to\ tilde {C}.

File: notebooks/4_SparseKernelMethods.ipynb (hunk 3 above)
Comment (bhelfrecht) [reply notebooks/4_SparseKernelMethods.ipynb:639]: Ah, yes, sorry. I missed the broader context when skimming for the change.

File: notebooks/5_CUR.ipynb (hunk 4 above)
Comment (bhelfrecht) [reply notebooks/5_CUR.ipynb]: Thank you for sending the other references and pointing this out -- your edit appears correct then. I'll double-check with @rosecers to make sure the PCov-CUR equation is made clearer.

**Review by bhelfrecht — APPROVED** [review]
